sklearn/utils/validation.py

```diff
@@ -1025,27 +1025,27 @@ def _check_sample_weight(sample_weight, X, dtype=None):
     """
     n_samples = _num_samples(X)
 
-    if dtype is not None and dtype not in [np.float32, np.float64]:
+    if hasattr(sample_weight, "dtype"):
+        dtype = sample_weight.dtype
+
+    if dtype not in [np.float32, np.float64]:
         dtype = np.float64
 
     if sample_weight is None or isinstance(sample_weight, numbers.Number):
         if sample_weight is None:
             sample_weight = np.ones(n_samples, dtype=dtype)
-        else:
+        elif isinstance(sample_weight, numbers.Number):
```

Comment:
the else statement was fine, wasn't it?

Reply:
I would even write

```python
if sample_weigt is None:
    sample_weight = np.ones(...)
elif isinstance(..., Number):
    ...
else:
    return sample_weight
```
Basically remove the first

Reply:
The `sample_weight` may have the wrong number of dimensions or elements if it's already an array. Thus, the further checks are necessary. We can only ignore the checks below if `sample_weight` is created within the function (i.e, when `sample_weight` is `None` or it's an `integer`